File: Benchmarks/PageRank/C++/optimized.cpp
#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

vector<string> splitString(const string &str, char delimiter) {
  stringstream stream(str);
  vector<string> split;
  string current;

  while (getline(stream, current, delimiter)) {
    split.push_back(current);
  }

  return split;
}

class Graph {
public:
  int vertices;
  vector<vector<bool>> incoming;
  vector<int> outgoing;

public:
  Graph(const string &path) {
    vector<vector<bool>> outgoing_edges;
    ifstream file(path);
    string line;

    while (getline(file, line)) {
      if (line[0] == 'p') {
        vector<string> values = splitString(line, ' ');
        int nvertices = atoi(values[2].c_str());

        this->vertices = nvertices;
        this->incoming = vector<vector<bool>>(nvertices);
        this->outgoing = vector<int>(nvertices);
        outgoing_edges = vector<vector<bool>>(nvertices);

        for (int i = 0; i < nvertices; i++) {
          this->incoming[i] = vector<bool>(nvertices);
          outgoing_edges[i] = vector<bool>(nvertices);
        }
      } else if (line[0] == 'a' || line[0] == 'e') {
        vector<string> values = splitString(line, ' ');

        int v0 = atoi(values[1].c_str());
        int v1 = atoi(values[2].c_str());

        outgoing_edges[v0][v1] = true;
        incoming[v1][v0] = true;
      }
    }

    for (int i = 0; i < this->vertices; i++) {
      this->outgoing[i] = std::count(outgoing_edges[i].begin(), outgoing_edges[i].end(), true);
    }

    file.close();
  }
};

float PageRank_single(int v, const Graph &graph, const vector<float> &ranks,
                      float damping) {
  float rank = (1 - damping) / graph.vertices;

  for (int u = 0; u < graph.vertices; u++) {
    if (graph.incoming[v][u]) {
      rank += damping * ranks[u] / graph.outgoing[u];
    }
  }

  return rank;
}

vector<float> PageRank(const Graph &graph, float damping,
                       float epsilon = 1e-4) {
  vector<float> ranks(graph.vertices);
  vector<float> newRanks(graph.vertices);

  for (int v = 0; v < graph.vertices; v++) {
    ranks[v] = 1.f / graph.vertices;
    newRanks[v] = 1.f / graph.vertices;
  }

  float change = epsilon + 1;
  while (change > epsilon) {
#pragma omp parallel for
    for (int v = 0; v < graph.vertices; v++) {
      newRanks[v] = PageRank_single(v, graph, ranks, damping);
    }

    change = 0.f;
#pragma omp parallel for
    for (int v = 0; v < graph.vertices; v++) {
      change += abs(ranks.at(v) - newRanks.at(v));
    }

    ranks = newRanks;
  }

  return ranks;
}

int main(int, const char **argv) {
  string file = argv[1];
  string out = argv[2];

  Graph graph(file);
  auto ranks = PageRank(graph, 0.85f);

  ofstream outFile(out);

  for (int v = 0; v < graph.vertices; v++) {
    outFile << v << ' ' << ranks[v] << '\n';
  }

  outFile.close();

  return 0;
}

File: Benchmarks/PageRank/C#/Optimized.cs
class Graph
{
    public System.Collections.Generic.HashSet<int>[] Incoming;
    public System.Collections.Generic.HashSet<int>[] Outgoing;
    public int NVertices = 0;

    public Graph(string path)
    {

        using (System.IO.StreamReader reader = new System.IO.StreamReader(path))
        {
            string line = reader.ReadLine();
            while (line != null)
            {
                if (line.StartsWith('p'))
                {
                    string[] values = line.Split(' ');
                    int vertices = int.Parse(values[2]);

                    NVertices = vertices;
                    Outgoing = new System.Collections.Generic.HashSet<int>[vertices];
                    Incoming = new System.Collections.Generic.HashSet<int>[vertices];

                    for (int i = 0; i < vertices; i++)
                    {
                        Outgoing[i] = new System.Collections.Generic.HashSet<int>();
                        Incoming[i] = new System.Collections.Generic.HashSet<int>();
                    }
                }
                else if (line.StartsWith('a') || line.StartsWith('e'))
                {
                    string[] values = line.Split(' ');
                    int a = int.Parse(values[1]);
                    int b = int.Parse(values[2]);

                    Outgoing[a].Add(b);
                    Incoming[b].Add(a);
                }

                line = reader.ReadLine();
            }
        }
    }
}

class PageRanker
{
    public static float PageRank_single(int v, Graph graph, float[] ranks, float damping)
    {
        float rank = (1.0f - damping) / graph.NVertices;

        foreach (int u in graph.Incoming[v])
            rank += damping * ranks[u] / graph.Outgoing[u].Count;

        return rank;
    }

    public static float[] PageRank(Graph graph, float damping, float epsilon = 1e-4f)
    {
        float[] ranks = new float[graph.NVertices];
        float[] newRanks = new float[graph.NVertices];

        for (int v = 0; v < graph.NVertices; v++)
        {
            ranks[v] = 1f / graph.NVertices;
            newRanks[v] = 1f / graph.NVertices;
        }


        float change = epsilon + 1f;
        while (change > epsilon)
        {
            System.Threading.Tasks.Parallel.For(0, graph.NVertices, v =>
            {
                newRanks[v] = PageRank_single(v, graph, ranks, damping);
            });

            // Recalculate change
            change = 0f;
            for (int v = 0; v < graph.NVertices; v++)
                change += System.MathF.Abs(ranks[v] - newRanks[v]);

            for (int v = 0; v < graph.NVertices; v++)
                ranks[v] = newRanks[v];
        }

        return ranks;
    }

    public static void Main(string[] args)
    {
        string file = args[0];
        string outFile = args[1];

        var graph = new Graph(file);
        var ranks = PageRank(graph, 0.85f);

        using (System.IO.StreamWriter writer = new System.IO.StreamWriter(outFile))
        {
            for (int v = 0; v < graph.NVertices; v++)
                writer.WriteLine($"{v} {ranks[v]}");
        }
    }
}